Sources/AblyChat/RoomLifecycleManager.swift

```diff
@@ -0,0 +1,297 @@
+import Ably
+
+/// The interface that the lifecycle manager expects its contributing realtime channels to conform to.
+///
+/// We use this instead of the ``RealtimeChannel`` interface as its ``attach`` and ``detach`` methods are `async` instead of using callbacks. This makes it easier to write mocks for (since ``RealtimeChannel`` doesn’t express to the type system that the callbacks it receives need to be `Sendable`, it’s hard to, for example, create a mock that creates a `Task` and then calls the callback from inside this task).
+///
+/// We choose to also mark the channel’s mutable state as `async`. This is a way of highlighting at the call site of accessing this state that, since `ARTRealtimeChannel` mutates this state on a separate thread, it’s possible for this state to have changed since the last time you checked it, or since the last time you performed an operation that might have mutated it, or since the last time you recieved an event informing you that it changed. To be clear, marking these as `async` doesn’t _solve_ these issues; it just makes them a bit more visible. We’ll decide how to address them in https://github.com/ably-labs/ably-chat-swift/issues/49.
+internal protocol RoomLifecycleContributorChannel: Sendable {
+    func attach() async throws(ARTErrorInfo)
+    func detach() async throws(ARTErrorInfo)
+
+    var state: ARTRealtimeChannelState { get async }
+    var errorReason: ARTErrorInfo? { get async }
+}
+
+internal actor RoomLifecycleManager<Channel: RoomLifecycleContributorChannel> {
+    /// A realtime channel that contributes to the room lifecycle.
+    internal struct Contributor {
+        /// The room feature that this contributor corresponds to. Used only for choosing which error to throw when a contributor operation fails.
+        internal var feature: RoomFeature
+
+        internal var channel: Channel
+    }
+
+    internal private(set) var current: RoomLifecycle
```

Comment:
bit nit picky/happy for you to disagree but I find this naming convention confusing... 

Properties with names like `current` or `shared` ring bells of us using singletons in my head - I know that's not the case here, but reading `RoomLifeCycleManager.current` doesn't immediately imply otherwise. 

Equally, I think `RoomLifecycle` is also a little misleading. To me a lifecycle encompasses all states, not just the current one. 

I'd prefer `RoomLifecycle` be renamed to `RoomState` or `RoomLifecycleState`, and `current` be renamed to `state`.

WDYT? 

(side note - `RoomStatus.current` which already exists makes perfect sense in that context)

Reply:
I'm happy for this to be addressed in a subsequent PR, but I think the discussion can happen here. It's easier to see where an API has shortcomings when seeing it in action as opposed to needing to revisit the original PR to have these discussions

Reply:
As discussed in standup, have split this into https://github.com/ably-labs/ably-chat-swift/issues/73